opensearch-observability/src/main/kotlin/org/opensearch/observability/model/Collaboration.kt

```diff
@@ -0,0 +1,319 @@
+
+            log.info(Thread.dumpStack())
+
+            while (XContentParser.Token.END_OBJECT != parser.nextToken()) {
+                val fieldName = parser.currentName()
+                parser.nextToken()
+                when (fieldName) {
+                    TAGS_TAG -> tags = parser.stringList()
+                    RESOLVED_TAG -> resolved = parser.booleanValue()
+                    else -> {
+                        val objectTypeForTag = CollaborationDataType.fromTagOrDefault(fieldName)
+                        if (objectTypeForTag != CollaborationDataType.NONE && typeInfo == null) {
+                            type = objectTypeForTag
+                            typeInfo = CollaborationTypeDataProperties.createObjectData(objectTypeForTag, parser)
+                        } else {
+                            parser.skipChildren()
+                            log.info("Unexpected field: $fieldName, while parsing CreateObservabilityObjectRequest")
+                        }
+                    }
+                }
+            }
+            parser.close()
+
+            log.info("====== Collaboration : start==========")
+            log.info("$TYPE_TAG: $type") // TODO: remove debug comments
+            log.info("$TAGS_TAG: $tags")
+            log.info("$RESOLVED_TAG: $resolved")
+            log.info("typeInfo : $typeInfo")
+            log.info("====== Collaboration : end ==========")
+            type ?: throw IllegalArgumentException("Collaboration type field absent")
+            // TODO: check wording, can this be empty? remove if needed
+            typeInfo ?: throw IllegalArgumentException("Collaboration type info field absent")
+
+            return Collaboration(type, tags, resolved, typeInfo)
+        }
+    }
+
+    /**
+     * create XContentBuilder from this object using [XContentFactory.jsonBuilder()]
+     * @param params XContent parameters
+     * @return created XContentBuilder object
+     */
+    fun toXContent(params: ToXContent.Params = ToXContent.EMPTY_PARAMS): XContentBuilder? {
+        return toXContent(XContentFactory.jsonBuilder(), params)
+    }
+
+    /**
+     * Constructor used in transport action communication.
+     * @param input StreamInput stream to deserialize data from.
+     */
+    constructor(input: StreamInput) : this(
+        type = input.readEnum(CollaborationDataType::class.java),
+        tags = input.readStringList(), // TODO change to list
+        resolved = input.readBoolean(),
+        typeInfo = input.readOptionalWriteable(getReaderForObjectType(input.readEnum(CollaborationDataType::class.java)))
+    )
+
+    /**
+     * {@inheritDoc}
+     */
+    override fun writeTo(output: StreamOutput) {
+        output.writeEnum(type)
+        output.writeStringCollection(tags)
+        output.writeBoolean(resolved)
+        output.writeOptionalWriteable(typeInfo)
+        output.writeEnum(type) // type is read twice in constructor
```

Comment:
type is read before typeinfo, this should be above reading typeinfo

Reply:
Done